### DISCUSSION TITLE
Incorporating Environment Variables?

### DISCUSSION THREAD

**Discussion by phantlantis** [spectreconsole/spectre.console#402] [Q&A] [answered]
Hi all-

Please forgive the n00b question. I was wondering peoples' thoughts on where/how to incorporate environment variables. Common rules of precedence dictates that cli params usually override environment variables, but when I attempt to add some options into the Constructor Initialization piece:

```csharp
public class CreateTicketSettings : CommandSettings
{
    public CreateTicketSettings(string platform)
    {
        var platformEV = System.Environment.GetEnvironmentVariable("PLATFORM");
        if (platform != null)
        {
            Platform = platform;
         }
        else Platform = platformEV;
    }
   
   [Description("Platform to create request for (web, ios, android)")]
   [CommandArgument(0, "<platform>")]
   public string Platform { get; set; }
}
```

 I can only ever successfully resolve the type (avoiding `"Error: Could not resolve type"`) if I set it to a required argument only with the angle brackets, but this fails validation because I didn't set the parameter in the CLI.  It seems like SetInterceptor could be an option, but I'm sure I'm overlooking something obvious.

**Comment by phil-scott-78** [comment]
I was trying to come up with something clever to hack around making this happen without resorting to uglifying the settings, but I'm coming up blank. I think it's a solid use case for adding support to the library.  But how to implement is something I'm not sure of.  I'd be tempted to go down one of two routes

1. Add a new attribute that would indicate that we should also look at a specific environment variable on the settings property itself. This way the user would opt in on a param-by-param basis and maybe even just try and grab system ones (e.g. `PATH`) if they wanted to go down that route.
2. Add a configuration option with an optional prefix that would enable this application wide.

If you get a chance can you add a new issue with the "Enhancement" tag? I'll move my thoughts over there too. Might be a better place for designing out this feature.

**Reply by phantlantis** [reply]
[Done.](https://github.com/spectreconsole/spectre.console/issues/405)
Thank you!

**Comment by patriksvensson** [comment, marked as answer]
Since you do not require the `Platform` parameter to be passed via the command line, it must be optional.

In this case, you can add custom validation that assigns the environment variable if `Platform` is not set or returns a validation error if neither is set.

```
public class CreateTicketSettings : CommandSettings
{
    [Description("Platform to create request for (web, ios, android)")]
    [CommandArgument(0, "[platform]")]
    public string Platform { get; set; }

    public override ValidationResult Validate()
    {
        // No platform provided?
        if (Platform == null)
        {
            // Use the one defined in the environment variable.
            var platformEV = System.Environment.GetEnvironmentVariable("PLATFORM");
            if (platformEV == null)
            {
                // Neither was set.
                return ValidationResult.Error("Platform has not been provided.");
            }
        }

        return ValidationResult.Success();
    }
}
```

I also have an idea of how to solve this more nicely, and I will create an issue and PR for it later today.

**Reply by phantlantis** [reply]
I'll mark this one as an answer since it does address the specific concerns I had, but I'm excited to see what the new approach is! I took the liberty of opening up an issue [here](https://github.com/spectreconsole/spectre.console/issues/405). Thanks for your help.

**Comment by koliyo** [comment]
See also PR in #540
